pkg/genericapiserver/openapi/openapi.go

```diff
@@ -0,0 +1,485 @@
+	// FixPath is a customization function that will be called on all generated path strings.
+	// This function is optional.
+	FixPath      func(path string) string
+	// FixResponses is a customization function that is called on all responses.
+	// This function is optional.
+	FixResponses func(responses *spec.Responses)
+	// List of webservice's path prefixes to ignore
+	IgnorePrefixes []string
+}
+
+type openAPI struct {
+	config       *Config
+	swagger      *spec.Swagger
+}
+
+// RegisterOpenAPIService registers a handler for /swagger.json and provides standard
+// swagger 2.0 (aka openAPI) specification.
+func RegisterOpenAPIService(config *Config, containers *restful.Container) (err error) {
+	o := openAPI{
+		config: config,
+	}
+	err = o.buildSwaggerSpec()
+	if err != nil {
+		return err
+	}
+	containers.ServeMux.HandleFunc("/swagger.json", func(w http.ResponseWriter, r *http.Request) {
+		resp := restful.NewResponse(w)
+		if r.URL.Path != "/swagger.json" {
+			resp.WriteErrorString(http.StatusNotFound, "Path not found!")
+		}
+		resp.WriteAsJson(o.swagger)
+	})
+	return nil
+}
+
+func (o *openAPI) buildSwaggerSpec() (err error) {
+	if o.swagger != nil {
+		return fmt.Errorf("Swagger spec is already built. Duplicate call to buildSwaggerSpec is not allowed.")
+	}
+	definitions, err := o.buildModels()
+	if err != nil {
+		return err
+	}
+	paths, err := o.buildPaths()
+	if err != nil {
+		return err
+	}
+	o.swagger = &spec.Swagger{
+		SwaggerProps: spec.SwaggerProps{
+			Swagger:     "2.0",
+			Definitions: definitions,
+			Paths:       &paths,
+			Info:        o.config.Info,
+		},
+	}
+	return nil
+}
+
+func (o *openAPI) buildModels() (definitions spec.Definitions, err error) {
+	definitions = spec.Definitions{}
+	for _, decl := range swagger.NewSwaggerBuilder(*o.config.SwaggerConfig).ProduceAllDeclarations() {
+		for _, oldNamedModel := range decl.Models.List {
+			_, ok := definitions[oldNamedModel.Name]
+			if ok {
+				// TODO(mbohlool): decide what to do with repeated models
+				continue
+			}
+			definitions[oldNamedModel.Name], err = buildModel(oldNamedModel.Model)
+			if err != nil {
+				return spec.Definitions{}, err
+			}
+		}
+	}
+	return definitions, nil
+}
+
+func buildModel(oldModel swagger.Model) (ret spec.Schema, err error) {
+	ret = spec.Schema{}
+	// model.SubTypes is not used in go-restful, ignoring.
+	ret.Description = oldModel.Description
+	ret.Required = oldModel.Required
+	ret.Discriminator = oldModel.Discriminator
+	ret.Properties = make(map[string]spec.Schema)
+	for _, oldProp := range oldModel.Properties.List {
+		if _, ok := ret.Properties[oldProp.Name]; ok {
+			return spec.Schema{}, fmt.Errorf("Duplicate property in old swagger spec.")
+		}
+		ret.Properties[oldProp.Name], err = buildProperty(oldProp)
+		if err != nil {
+			return ret, err
+		}
+	}
+	return ret, nil
+}
+
+func buildProperty(oldProperty swagger.NamedModelProperty) (newProperty spec.Schema, err error) {
```

Comment:
similar comment as above. I think calling them swaggerProperty and openAPIProperty is better than calling them old and new property.